File: src/utils/telemetry.ts
// Self-owned crash telemetry — NO third-party service.
//
// Replaces the previous PostHog analytics + Sentry error tracking (both were
// third-party systems that sent data to infrastructure we don't control). Per
// the project owner's requirement, everything now flows to an endpoint THEY
// own (a Google Apps Script that writes to their own Google Sheet — see
// docs/crash-telemetry-setup.md). The actual HTTP POST happens in the Rust
// backend (commands/telemetry.rs) so the endpoint URL never lives in the DOM
// and there are no CORS constraints.
//
// This module is only the frontend capture layer: it installs global error
// handlers and forwards crashes to the backend WHEN the user has opted in.
// Nothing is ever sent while `_enabled` is false — the default.

import { invoke } from "@tauri-apps/api/core";

let _enabled = false;
let _installed = false;

// Anonymous, random, generated once and kept in localStorage. NOT tied to any
// identity — its only purpose is to let us tell "one machine crashed 20 times"
// apart from "20 machines crashed once", which changes how we triage. It can
// be wiped by clearing app data and never leaves the user's control.
const _installId = (() => {
  const KEY = "nxs_install_id";
  let id = localStorage.getItem(KEY);
  if (!id) {
    id = crypto.randomUUID();
    localStorage.setItem(KEY, id);
  }
  return id;
})();

interface CrashPayload {
  kind: string;        // "js-error" | "unhandled-rejection"
  message: string;
  stack: string;
  url: string;         // minified bundle location, e.g. "index-abc.js:102" — not sensitive
}

// Strip OS filesystem paths out of anything we send. Backend errors reaching the
// frontend as rejected `invoke` promises routinely interpolate real paths — and
// even vault file NAMES ("'budget.xlsx' is not in this vault") — so an unhandled
// rejection could otherwise ship a user's private path/filename to the crash
// sheet, breaking the "we never send file paths" promise (and the vault's).
// We target Windows drive paths (`C:\…`) and UNC paths (`\\server\…`); minified
// bundle locations use forward-slash URLs (`…/index-abc.js:102`) and are left
// intact so stacks stay useful.
function scrubPaths(s: string): string {
  return s
    .replace(/[A-Za-z]:\\[^\s"'<>|]*/g, "<path>")
    .replace(/\\\\[^\s"'<>|]+/g, "<path>");
}

function send(p: CrashPayload) {
  if (!_enabled) return;
  // Fire-and-forget. The backend buffers to disk on network failure and
  // retries at next launch, so a dropped promise here is not a lost report.
  invoke("submit_crash_report", {
    report: {
      kind: p.kind,
      message: scrubPaths(p.message).slice(0, 2000),
      stack: scrubPaths(p.stack).slice(0, 8000),
      url: p.url.slice(0, 500),
      install_id: _installId,
      app_version: __APP_VERSION__,
    },
  }).catch(() => {});
}

// Installed exactly once, regardless of how many times telemetryInit runs
// (React strict mode, settings reloads). The handlers themselves always
// check `_enabled` at fire time, so toggling the setting live takes effect
// without re-registering.
function installHandlers() {
  if (_installed) return;
  _installed = true;

  window.addEventListener("error", (e) => {
    // Filter noise that isn't actionable and floods reports.
    const msg = e.message ?? "";
    if (msg.includes("ResizeObserver loop")) return;
    send({
      kind: "js-error",
      message: msg || String(e.error ?? "unknown"),
      stack: e.error?.stack ?? "",
      url: `${e.filename ?? ""}:${e.lineno ?? 0}:${e.colno ?? 0}`,
    });
  });

  window.addEventListener("unhandledrejection", (e) => {
    const reason = e.reason;
    send({
      kind: "unhandled-rejection",
      message: reason?.message ?? String(reason ?? "unknown"),
      stack: reason?.stack ?? "",
      url: "",
    });
  });
}

/// Called once at startup with the persisted opt-in value. Installs the global
/// handlers (they're cheap and always gated) and, when enabled, asks the
/// backend to flush any crash reports buffered while offline / from a previous
/// Rust panic.
export function telemetryInit(enabled: boolean) {
  _enabled = enabled;
  installHandlers();
  // Sync the backend gate (the Rust panic hook reads it) BEFORE flushing.
  invoke("set_crash_reporting_enabled", { enabled }).catch(() => {});
  if (enabled) {
    invoke("flush_pending_crash_reports").catch(() => {});
  }
}

/// Called when the user toggles the setting so the change takes effect live —
/// both the JS-side gate and the Rust-side gate move together.
export function telemetrySetEnabled(enabled: boolean) {
  _enabled = enabled;
  invoke("set_crash_reporting_enabled", { enabled }).catch(() => {});
  if (enabled) invoke("flush_pending_crash_reports").catch(() => {});
}

File: src-tauri/src/commands/telemetry.rs
//! Self-owned crash telemetry — sends crash reports to an endpoint the project
//! owner controls (a Google Apps Script that writes to their own Google Sheet;
//! see docs/crash-telemetry-setup.md). NO third-party telemetry service.
//!
//! ## Flow
//!
//! - Frontend catches JS errors / unhandled rejections and calls
//!   `submit_crash_report` (only when the user opted in).
//! - Rust panics are appended to a local buffer file by the panic hook in
//!   `lib.rs` (see `buffer_panic`).
//! - `submit_crash_report` POSTs immediately; on network failure it appends to
//!   the buffer. `flush_pending_crash_reports` (called at startup and on opt-in)
//!   drains the buffer.
//!
//! ## Privacy
//!
//! We send: error kind, message, stack, minified bundle location, an anonymous
//! random install id, app version, OS, arch, timestamp. We NEVER send file
//! paths (beyond minified JS locations like `index-abc.js:102`), file contents,
//! tags, AI conversations, or the DB. Nothing is sent unless the user turned
//! crash reporting on (default off).

use serde::{Deserialize, Serialize};
use std::io::Write;
use std::path::PathBuf;
use std::sync::atomic::{AtomicBool, Ordering};

// ── Configuration (fill these in before crash reporting works) ───────────────
//
// ENDPOINT: the Apps Script Web App URL (…/exec) from your deployment.
// SHARED_SECRET: a random string you also paste into the Apps Script. It's not
// a true secret (it ships in the binary and can be extracted), but it stops
// casual spam POSTs to your endpoint. Rotate it in both places if abused.
//
// While ENDPOINT is left at the placeholder, submit/flush are silent no-ops —
// the app behaves exactly as if crash reporting were off, so shipping without
// configuring the endpoint is safe.
const ENDPOINT: &str = "https://script.google.com/macros/s/AKfycbyVd2NWtlJJ9mDELNLZSLz-cW35pErq_kS9vgZc50ccvlZP94f0K5rSrGD_2C91d1b-Kw/exec";
const SHARED_SECRET: &str = "ioehfboibdsofvkkjb54154!";

/// Keep at most this many buffered reports on disk. Bounds the file if the
/// endpoint is unreachable for a long time; oldest are dropped first.
const MAX_BUFFERED: usize = 50;

/// Process-wide opt-in flag. Mirrors the `telemetryEnabled` setting. Read by
/// the panic hook (which can't touch the DB while the app is dying) so panics
/// are only buffered when the user consented.
static ENABLED: AtomicBool = AtomicBool::new(false);

pub fn is_enabled() -> bool {
    ENABLED.load(Ordering::Relaxed)
}

/// The crash report as sent by the frontend + enriched server-side.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CrashReport {
    pub kind: String,
    pub message: String,
    pub stack: String,
    pub url: String,
    pub install_id: String,
    pub app_version: String,
    // Enriched in `enrich` before sending; frontend leaves these empty.
    #[serde(default)]
    pub os: String,
    #[serde(default)]
    pub arch: String,
    #[serde(default)]
    pub ts: i64,
}

fn enrich(mut r: CrashReport) -> CrashReport {
    if r.os.is_empty() {
        r.os = std::env::consts::OS.to_string();
    }
    if r.arch.is_empty() {
        r.arch = std::env::consts::ARCH.to_string();
    }
    if r.ts == 0 {
        r.ts = std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs() as i64)
            .unwrap_or(0);
    }
    r
}

/// Local buffer path: `%LOCALAPPDATA%\com.nxs.app\pending-crashes.jsonl`.
/// Computed from the env (not a Tauri AppHandle) so the panic hook — which
/// runs without easy handle access while the app is dying — can reach it too.
/// One JSON object per line (JSONL) for append-only, partial-read safety.
pub fn buffer_path() -> Option<PathBuf> {
    let base = std::env::var_os("LOCALAPPDATA")?;
    Some(PathBuf::from(base).join("com.nxs.app").join("pending-crashes.jsonl"))
}

fn append_to_buffer(report: &CrashReport) {
    let Some(path) = buffer_path() else { return };
    if let Some(parent) = path.parent() {
        let _ = std::fs::create_dir_all(parent);
    }
    let Ok(line) = serde_json::to_string(report) else { return };
    if let Ok(mut f) = std::fs::OpenOptions::new().create(true).append(true).open(&path) {
        let _ = writeln!(f, "{}", line);
    }
    trim_buffer(&path);
}

/// Keep only the last MAX_BUFFERED lines so a long offline stretch can't grow
/// the file without bound.
fn trim_buffer(path: &PathBuf) {
    let Ok(content) = std::fs::read_to_string(path) else { return };
    let lines: Vec<&str> = content.lines().filter(|l| !l.trim().is_empty()).collect();
    if lines.len() <= MAX_BUFFERED {
        return;
    }
    let kept = &lines[lines.len() - MAX_BUFFERED..];
    let _ = std::fs::write(path, kept.join("\n") + "\n");
}

/// POST a single report. Returns Ok(()) on 2xx, Err otherwise. Placeholder
/// endpoint → treated as success no-op (nothing to send to).
async fn post_report(client: &reqwest::Client, report: &CrashReport) -> Result<(), String> {
    if ENDPOINT.starts_with("REPLACE_WITH_") {
        return Ok(());
    }
    let body = serde_json::json!({
        "type": "crash",
        "secret": SHARED_SECRET,
        "report": report,
    });
    let resp = client
        .post(ENDPOINT)
        .json(&body)
        .send()
        .await
        .map_err(|e| e.to_string())?;
    if resp.status().is_success() {
        Ok(())
    } else {
        Err(format!("HTTP {}", resp.status()))
    }
}

fn http_client() -> reqwest::Client {
    reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(15))
        .user_agent("nxs-crash-reporter")
        .build()
        .unwrap_or_default()
}

// ── Tauri commands ───────────────────────────────────────────────────────────

/// Called by the frontend at startup and when the user toggles the setting.
/// Syncs the process-wide flag the panic hook reads.
#[tauri::command]
pub fn set_crash_reporting_enabled(enabled: bool) {
    ENABLED.store(enabled, Ordering::Relaxed);
}

/// Submit one crash report. Sends immediately; on failure, buffers to disk for
/// a later flush. Silently no-ops when disabled or endpoint unconfigured, so
/// the frontend never has to care about either condition.
#[tauri::command]
pub async fn submit_crash_report(report: CrashReport) -> Result<(), String> {
    if !is_enabled() {
        return Ok(());
    }
    let report = enrich(report);
    let client = http_client();
    if post_report(&client, &report).await.is_err() {
        append_to_buffer(&report);
    }
    Ok(())
}

/// Drain the on-disk buffer. Reports that still fail to send are kept for the
/// next attempt. No-op when disabled.
#[tauri::command]
pub async fn flush_pending_crash_reports() -> Result<(), String> {
    if !is_enabled() {
        return Ok(());
    }
    let Some(path) = buffer_path() else { return Ok(()) };
    let content = match std::fs::read_to_string(&path) {
        Ok(c) => c,
        Err(_) => return Ok(()), // no buffer file = nothing to flush
    };
    let reports: Vec<CrashReport> = content
        .lines()
        .filter(|l| !l.trim().is_empty())
        .filter_map(|l| serde_json::from_str::<CrashReport>(l).ok())
        .collect();
    if reports.is_empty() {
        let _ = std::fs::remove_file(&path);
        return Ok(());
    }

    let client = http_client();
    let mut still_failing: Vec<CrashReport> = Vec::new();
    for report in reports {
        if post_report(&client, &report).await.is_err() {
            still_failing.push(report);
        }
    }

    if still_failing.is_empty() {
        let _ = std::fs::remove_file(&path);
    } else {
        // Rewrite with only the ones that still didn't make it.
        let lines: Vec<String> = still_failing
            .iter()
            .filter_map(|r| serde_json::to_string(r).ok())
            .collect();
        let _ = std::fs::write(&path, lines.join("\n") + "\n");
    }
    Ok(())
}

/// Called by the panic hook in lib.rs. Best-effort, synchronous, no HTTP (the
/// app may be dying) — just append to the buffer for the next launch to flush.
/// Gated on the process-wide flag so a user who never opted in accumulates
/// nothing.
pub fn buffer_panic(message: &str, location: &str) {
    if !is_enabled() {
        return;
    }
    let report = enrich(CrashReport {
        kind: "rust-panic".to_string(),
        message: message.chars().take(2000).collect(),
        stack: String::new(),
        url: location.chars().take(500).collect(),
        // No frontend install id available in the panic hook. The flush still
        // sends it; the admin view shows an empty install id for panics, which
        // is acceptable (they're rarer and the message+location identify them).
        install_id: String::new(),
        app_version: env!("CARGO_PKG_VERSION").to_string(),
        os: String::new(),
        arch: String::new(),
        ts: 0,
    });
    append_to_buffer(&report);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn enrich_fills_missing_fields() {
        let r = enrich(CrashReport {
            kind: "js-error".into(),
            message: "boom".into(),
            stack: "".into(),
            url: "".into(),
            install_id: "abc".into(),
            app_version: "0.1.0".into(),
            os: "".into(),
            arch: "".into(),
            ts: 0,
        });
        assert!(!r.os.is_empty());
        assert!(!r.arch.is_empty());
        assert!(r.ts > 0);
    }

    #[test]
    fn enrich_preserves_provided_fields() {
        let r = enrich(CrashReport {
            kind: "js-error".into(),
            message: "boom".into(),
            stack: "".into(),
            url: "".into(),
            install_id: "abc".into(),
            app_version: "0.1.0".into(),
            os: "customos".into(),
            arch: "customarch".into(),
            ts: 12345,
        });
        assert_eq!(r.os, "customos");
        assert_eq!(r.arch, "customarch");
        assert_eq!(r.ts, 12345);
    }

    #[test]
    fn disabled_by_default() {
        // A fresh process must not have crash reporting on.
        assert!(!is_enabled());
    }
}
